numba/npyufunc/workqueue.c

```diff
@@ -0,0 +1,249 @@
+/*
+Implement parallel vectorize workqueue.
+
+This keeps a set of worker threads running all the time.
+They wait and spin on a task queue for jobs.
+
+**WARNING**
+This module is not thread-safe.  Adding task to queue is not protected from
+race condition.
+*/
+
+#ifdef _MSC_VER
+    /* Windows */
+    #include <windows.h>
+	#include <process.h>
+    #define NUMBA_WINTHREAD
+#else
+    /* PThread */
+    #include <pthread.h>
+    #include <unistd.h>
+    #define NUMBA_PTHREAD
+#endif
+
+#include <string.h>
+#include <stdio.h>
+#include "workqueue.h"
+#include "../_pymodule.h"
+
+static cas_function_t *cas = NULL;
+
+void cas_wait(volatile int *ptr, const int old, const int repl) {
+    int out = repl;
+    int timeout = 1;   /* starting from 1us nap */
+    static const int MAX_WAIT_TIME = 20 * 1000; /* max wait is 20ms */
+
+    while (1) {
+        if (cas) { /* protect against CAS function being released by LLVM during
+                      interpreter teardown. */
+            out = cas(ptr, old, repl);
+            if (out == old) return;
+        }
+
+        take_a_nap(timeout);
+
+        /* Exponentially increase the wait time until the max has reached*/
+        timeout <<= 1;
+        if (timeout >= MAX_WAIT_TIME) {
+            timeout = MAX_WAIT_TIME;
+        }
+    }
+}
+
+/* As the thread-pool isn't inherited by children,
+   free the task-queue, too. */
+static void reset_after_fork(void);
+
+/* PThread */
+#ifdef NUMBA_PTHREAD
+
+thread_pointer numba_new_thread(void *worker, void *arg)
+{
+    int status;
+    pthread_attr_t attr;
+    pthread_t th;
+
+    pthread_atfork(0, 0, reset_after_fork);
```

Comment:
This will register one new callback for each thread launched, which sounds wasteful.